File: PythonPlugins/CountryBlackList/CountryBlackList.py
__title__ = 'CountryBlackList'
__author__ = 'Jakkee'
__version__ = '2.0.1'

try:
    import clr
    clr.AddReferenceByPartialName("Fougerite", "GeoIP")
    import Fougerite
    import GeoIP
    from GeoIP import GeoIP as RealGeoIP  
except:
    raise ImportError("Failed to reference the GeoIP.dll, Download from: http://fougerite.com/resources/geoip.135/")

geo = RealGeoIP.Instance
PluginSettings = {}


class CountryBlackList:
    def On_PluginInit(self):
        Util.ConsoleLog(__title__ + " by " + __author__ + " Version: " + __version__ + " loaded.", False)
        if not Plugin.IniExists("Settings"):
            Plugin.CreateIni("Settings")
            ini = Plugin.GetIni("Settings")
            ini.AddSetting("Settings", "ShowAcceptedMessage", "true")
            ini.AddSetting("Settings", "ShowDeniedMessage", "true")
            ini.AddSetting("Settings", "LogTimedOutConnection", "true")
            ini.AddSetting("Settings", "EnableWhitelist", "true")
            ini.AddSetting("BlackList", "BlackListedCountries", "TK, JO")
            ini.AddSetting("Messages", "JoinMessage", "%PLAYER% has connected from: %COUNTRY%")
            ini.AddSetting("Messages", "PlayerDisconnectMessage", "Your country is on the servers blacklist")
            ini.AddSetting("Messages", "ServerDisconnectMessage", "%PLAYER% is trying to connect from: %COUNTRY% but is black listed")
            ini.AddSetting("Messages", "UnknownLocation", "A hidden location")
            ini.Save()
        if not Plugin.IniExists("WhiteList"):
            Plugin.CreateIni("WhiteList")
            ini = Plugin.GetIni("WhiteList")
            ini.AddSetting("SteamID", "IP Address", "Players Name")
            ini.AddSetting("76561198135558142", "127.0.0.1", "Xiled Jakkee")
            ini.Save()
        PluginSettings.clear()
        ini = Plugin.GetIni("Settings")
        PluginSettings["SAM"] = ini.GetBoolSetting("Settings", "ShowAcceptedMessage")
        PluginSettings["SDM"] = ini.GetBoolSetting("Settings", "ShowDeniedMessage")
        PluginSettings["LTOC"] = ini.GetBoolSetting("Settings", "LogTimedOutConnection")
        PluginSettings["EW"] = ini.GetBoolSetting("Settings", "EnableWhitelist")
        PluginSettings["BL"] = ini.GetSetting("BlackList", "BlackListedCountries")
        PluginSettings["JM"] = ini.GetSetting("Messages", "JoinMessage")
        PluginSettings["PDM"] = ini.GetSetting("Messages", "PlayerDisconnectMessage")
        PluginSettings["DM"] = ini.GetSetting("Messages", "ServerDisconnectMessage")
        PluginSettings["UL"] = ini.GetSetting("Messages", "UnknownLocation")

    def find(self, blacklist, CountryCode):
        try:
            blacklist = blacklist.Replace(" ", "")
            blacklist = blacklist.split(',')
            for configlist in blacklist:
                if CountryCode == configlist:
                    return True
                else:
                    continue
            return False
        except:
            return False

    def OnWhiteList(self, Player):
        if Player.Admin or Player.Moderator:
            return True
        else:
            if Plugin.IniExists("WhiteList"):
                ini = Plugin.GetIni("WhiteList")
                for id in ini.Sections:
                    if ini.ContainsSetting(id, Player.IP):
                        return True
                    if id == Player.SteamID:
                        return True
            return False

    def On_Command(self, Player, cmd, args):
        if cmd == "wlist":
            if Player.Admin or Player.Moderator:
                if len(args) == 0:
                    Player.MessageFrom("CountryBlackList", "Usage: /wlist <User Name>")
                    Player.MessageFrom("CountryBlackList", "Adds a user to the whitelist")
                else:
                    target = self.CheckV(Player, args)
                    if target is not None:
                        if not Plugin.IniExists("WhiteList"):
                            Plugin.CreateIni("WhiteList")
                        ini = Plugin.GetIni("WhiteList")
                        ini.AddSetting(target.SteamID, target.IP, target.Name)
                        ini.Save()
            else:
                Player.MessageFrom("CountryBlackList", "You are not allowed to use this command!")

    def On_PlayerConnected(self, Player):
        try:                    
            IPData = geo.GetDataOfIP(Player.IP)
            if IPData is None:
                if PluginSettings["SAM"]:
                    msg = PluginSettings["JM"]
                    msg = msg.Replace("%PLAYER%", Player.Name)
                    msg = msg.Replace("%COUNTRY%", PluginSettings["UL"])
                    Server.Broadcast(msg)
                if PluginSettings["LTOC"]:
                    Plugin.Log("Unknown Location", Player.SteamID + "=" + "Player.Name" + " [" + Player.IP + "]")
                return
            countrycode = IPData.CountryShort
            blacklist = PluginSettings["BL"]
            if self.find(blacklist, countrycode):
                if PluginSettings["EW"]:
                    if OnWhiteList(Player.SteamID, Player.IP):
                        if PluginSettings["SAM"]:
                            msg = PluginSettings["JM"]
                            msg = msg.Replace("%PLAYER%", Player.Name)
                            msg = msg.Replace("%COUNTRY%", IPData.Country)
                            Server.Broadcast(msg)
                        return
                playerdisconnectMSG = PluginSettings["PDM"]
                if PluginSettings["SDM"]:
                    msg = PluginSettings["DM"]
                    msg = msg.Replace("%PLAYER%", Player.Name)
                    msg = msg.Replace("%COUNTRY%", IPData.Country)
                    Server.Broadcast(msg)
                Player.Message(playerdisconnectMSG + " [" + countrycode + "]")
                Player.Disconnect()
            else:
                if PluginSettings["SAM"]:
                    msg = PluginSettings["JM"]
                    msg = msg.Replace("%PLAYER%", Player.Name)
                    msg = msg.Replace("%COUNTRY%", IPData.Country)
                    Server.Broadcast(msg)
        except:
            if PluginSettings["SAM"]:
                msg = PluginSettings["JM"]
                msg = msg.Replace("%PLAYER%", Player.Name)
                msg = msg.Replace("%COUNTRY%", PluginSettings["UL"])
                Server.Broadcast(msg)
            if PluginSettings["LTOC"]:
                Plugin.Log("Unknown Location", Player.SteamID + "=" + "Player.Name" + " [" + Player.IP + "]")

#DreTaX's amazing shit here...
    def GetPlayerName(self, namee):
        try:
            name = namee.lower()
            for pl in Server.Players:
                if pl.Name.lower() == name:
                    return pl
            return None
        except:
            return None

#Also here...
    def CheckV(self, Player, args):
        count = 0
        if hasattr(args, '__len__') and (not isinstance(args, str)):
            p = self.GetPlayerName(str.join(" ", args))
            if p is not None:
                return p
            for pl in Server.Players:
                for namePart in args:
                    if namePart.lower() in pl.Name.lower():
                        p = pl
                        count += 1
                        continue
        else:
            nargs = str(args).lower()
            p = self.GetPlayerName(nargs)
            if p is not None:
                return p
            for pl in Server.Players:
                if nargs in pl.Name.lower():
                    p = pl
                    count += 1
                    continue
        if count == 0:
            Player.Message("Couldn't find " + str.join(" ", args) + "!")
            return None
        elif count == 1 and p is not None:
            return p
        else:
            Player.Message("Found " + str(count) + " players with similar name!")
            Player.Message("Use quotation marks if the name has a space in it.")

File: CSharpPlugins/CountryBlackList/CountryBlackList.cs
using Fougerite;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using UnityEngine;

namespace CountryBlackList
{
    public class CountryBlackList : Fougerite.Module
    {
        internal bool Show_Accept_Message = true;
        internal bool Show_Denied_Message = true;
        internal bool Use_WhiteList = true;
        internal bool Log_TimedOut_Connections = true;
        internal string Join_Message = "%PLAYER% has connected from %COUNTRY%";
        internal string Player_Disconnect_Message = "%COUNTRY% is on the servers blacklist";
        internal string Server_Disconnect_Message = "%PLAYER% has connected from %COUNTRY% but is blacklisted";
        internal string Unknown_Location = "Unknown Location";
        internal List<string> BlackList = new List<string> { "TK", "JO" };
        internal IniParser Settings;
        internal IniParser WhiteList;
        internal System.IO.StreamWriter file;
        private static CountryBlackList _inst;
        private GameObject GameO;
        private CountryBlackListMono Handler;

        public override string Name
        {
            get
            {
                return "CountryBlackList";
            }
        }

        public override string Author
        {
            get
            {
                return "Jakkee & DreTaX";
            }
        }

        public override string Description
        {
            get
            {
                return "Block any country you wish";
            }
        }

        public override Version Version
        {
            get
            {
                return new Version("3.1.0");
            }
        }

        public static CountryBlackList Instance
        {
            get
            {
                return _inst;
            }
        }

        public override void Initialize()
        {
            _inst = this;
            Hooks.OnCommand += new Hooks.CommandHandlerDelegate(On_Command);
            Hooks.OnPlayerConnected += new Hooks.ConnectionHandlerDelegate(On_PlayerConnected);
            ReloadConfig();

            GameO = new GameObject();
            Handler = GameO.AddComponent<CountryBlackListMono>();
            UnityEngine.Object.DontDestroyOnLoad(GameO);
        }

        public override void DeInitialize()
        {
            Hooks.OnCommand -= new Hooks.CommandHandlerDelegate(On_Command);
            Hooks.OnPlayerConnected -= new Hooks.ConnectionHandlerDelegate(On_PlayerConnected);
            if (GameO != null)
            {
                UnityEngine.Object.Destroy(GameO);
                GameO = null;
            }
        }

        private void ReloadConfig()
        {
            if (!File.Exists(Path.Combine(ModuleFolder, "Settings.ini")))
            {
                File.Create(Path.Combine(ModuleFolder, "Settings.ini")).Dispose();
                Settings = new IniParser(Path.Combine(ModuleFolder, "Settings.ini"));
                Settings.AddSetting("Settings", "Show Accepted Message", "True");
                Settings.AddSetting("Settings", "Show Denied Message", "True");
                Settings.AddSetting("Settings", "Use WhiteList", "True");
                Settings.AddSetting("Settings", "Log TimedOut Connections", "True");
                Settings.AddSetting("BlackList", "BlackListed Countries", "TK, JO");
                Settings.AddSetting("Messages", "Join Message", "%PLAYER% has connected from %COUNTRY%");
                Settings.AddSetting("Messages", "Unknown Location", "Unknown Location");
                Settings.AddSetting("Messages", "Player Disconnect Message", "%COUNTRY% is on the servers blacklist");
                Settings.AddSetting("Messages", "Server Disconnect Message", "%PLAYER% has connected from %COUNTRY% but is blacklisted");
                Settings.Save();
            }
            Settings = new IniParser(Path.Combine(ModuleFolder, "Settings.ini"));

            try
            {
                Show_Accept_Message = Boolean.Parse(Settings.GetSetting("Settings", "Show Accepted Message"));
                Show_Denied_Message = Boolean.Parse(Settings.GetSetting("Settings", "Show Denied Message"));
                Use_WhiteList = Boolean.Parse(Settings.GetSetting("Settings", "Use WhiteList"));
                Log_TimedOut_Connections = Boolean.Parse(Settings.GetSetting("Settings", "Log TimedOut Connections"));
                BlackList = Settings.GetSetting("BlackList", "BlackListed Countries").Replace(" ", "").Split(',').ToList();
                Join_Message = Settings.GetSetting("Messages", "Join Message");
                Player_Disconnect_Message = Settings.GetSetting("Messages", "Player Disconnect Message");
                Server_Disconnect_Message = Settings.GetSetting("Messages", "Server Disconnect Message");
                Unknown_Location = Settings.GetSetting("Messages", "Unknown Location");
            }
            catch
            {
                Logger.LogError("[CountryBlackList] Error in settings.ini, Check your settings & reload! Using default settings for now");
                Show_Accept_Message = true;
                Show_Denied_Message = true;
                Use_WhiteList = true;
                Log_TimedOut_Connections = true;
                Join_Message = "%PLAYER% has connected from %COUNTRY%";
                Player_Disconnect_Message = "%COUNTRY% is on the servers blacklist";
                Server_Disconnect_Message = "%PLAYER% has connected from %COUNTRY% but is blacklisted";
                Unknown_Location = "Unknown Location";
                BlackList = "TK, JO".Replace(" ", "").Split(',').ToList();
            }

            if (!File.Exists(Path.Combine(ModuleFolder, "WhiteList.ini")))
            {
                File.Create(Path.Combine(ModuleFolder, "WhiteList.ini")).Dispose();
                WhiteList = new IniParser(Path.Combine(ModuleFolder, "WhiteList.ini"));
                WhiteList.AddSetting("SteamID", "IP Address", "Players Name");
                WhiteList.AddSetting("76561198135558142", "127.0.0.1", "Xiled Jakkee");
                WhiteList.Save();
            }
            WhiteList = new IniParser(Path.Combine(ModuleFolder, "WhiteList.ini"));
        }

        public void On_Command(Fougerite.Player player, string cmd, string[] args)
        {
            if (cmd == "wlist")
            {
                if (player.Admin || player.Moderator)
                {
                    if (args.Length.Equals(0))
                    {
                        player.MessageFrom("CountryBlackList", "Usage: /wlist <User Name>");
                        player.MessageFrom("CountryBlackList", "Adds a user to the whitelist");
                    }
                    else
                    {
                        string search = string.Join(" ", args);
                        Fougerite.Player target = Fougerite.Server.GetServer().FindPlayer(search);
                        if (target != null)
                        {
                            if (!OnWhiteList(player.SteamID, player.IP))
                            {
                                WhiteList.AddSetting(target.SteamID, target.IP, target.Name);
                                WhiteList.Save();
                                player.MessageFrom("CountryBlackList", "You have added " + target.Name + " to the whitelist");
                            }
                            else
                            {
                                player.MessageFrom("CountryBlackList", target.Name + " is already on the whitelist!");
                            }
                        }
                    }
                }
                else
                {
                    player.MessageFrom("CountryBlackList", "You are not allowed to use this command!");
                }
            }
            else if (cmd == "dlist")
            {
                if (player.Admin || player.Moderator)
                {
                    if (args.Length.Equals(0))
                    {
                        player.MessageFrom("CountryBlackList", "Usage: /dlist <User Name>");
                        player.MessageFrom("CountryBlackList", "Removes a user from the whitelist");
                    }
                    else
                    {
                        string search = string.Join(" ", args);
                        Fougerite.Player target = Fougerite.Server.GetServer().FindPlayer(search);
                        if (target != null)
                        {
                            if (OnWhiteList(player.SteamID, player.IP))
                            {
                                WhiteList.DeleteSetting(target.SteamID, target.IP);
                                WhiteList.Save();
                                player.MessageFrom("CountryBlackList", "You have removed " + target.Name + " from the whitelist");
                            }
                            else
                            {
                                player.MessageFrom("CountryBlackList", target.Name + " is not on the whitelist!");
                            }
                        }
                    }
                }
                else
                {
                    player.MessageFrom("CountryBlackList", "You are not allowed to use this command!");
                }
            }
            else if (cmd == "cbreload")
            {
                if (player.Admin)
                {
                    player.MessageFrom("CountryBlackList", "Reloading config...");
                    ReloadConfig();
                    player.MessageFrom("CountryBlackList", "Done");
                }
            }
        }

        public bool OnWhiteList(string playerid, string playerip)
        {
            foreach (var steamid in WhiteList.Sections)
            {
                if (WhiteList.ContainsSetting(steamid, playerip))
                {
                    return true;
                }
                if (steamid == playerid)
                {
                    return true;
                }
            }
            return false;
        }

        public void On_PlayerConnected(Fougerite.Player player)
        {
            Handler.HandleGeoIPRequest(player);
        }
    }
}